README.md
# Distortion Effect Carousel

HTML carousel with distortion effect for React and Vanilla Javascript.\
Inspired by [hover-effect](https://github.com/robin-dela/hover-effect).

[DEMO](https://shasaf1986.github.io/distortion-effect-carousel)

![Build CI](https://github.com/shasaf1986/distortion-effect-carousel/workflows/Build%20CI/badge.svg)
![npm](https://img.shields.io/npm/v/distortion-effect-carousel)

### Install

With yarn.

```
yarn add distortion-effect-carousel
```

With NPM.

```
npm install distortion-effect-carousel
```

## Basic usage

```jsx
import React from 'react';
import { useDistortionEffectCarousel } from 'distortion-effect-carousel';

interface MyCarouselProps {
  images: string[];
  displacmentImage: string;
}

const MyCarousel: React.FC<MyCarouselProps> = ({
  displacmentImage,
  images,
}) => {
  const { ref, next, prev } = useDistortionEffectCarousel({
    images,
    displacmentImage,
  });

  // ... do something with next() and prev()
  return (
    <div
      style={{
        height: '350px',
      }}
      ref={ref}
    />
  );
};
```

## Hook

### Parameters

| Name                         | Type                                                       | Default                   | Description                                                                                        |
| ---------------------------- | ---------------------------------------------------------- | ------------------------- | -------------------------------------------------------------------------------------------------- |
| `images`                     | `string[]`                                                 |                           | Array of image sources                                                                             |
| `displacementImage`          | `string`                                                   |                           | The source of `displacementImage` used to do the transition between two images.                    |
| `ref?`                       | `MutableRefObject<T extends HTMLElement = HTMLDivElement>` |                           | The ref of the parent DOM element                                                                  |
| `intensity?`                 | `number`                                                   | `1`                       | Used to determine the intensity of the distortion effect. 0 is no effect and 1 is full distortion. |
| `commonAngle?`               | `number`                                                   | `Math.PI / 4`             | Angle of the distortion effect in Radians. Defaults to Pi / 4 (45 degrees).                        |
| `angle1?`                    | `number`                                                   | `commonAngle`             | Overrides the distortion angle for the first image.                                                |
| `angle2?`                    | `number`                                                   | `-commonAngle * 3`        | Overrides the distortion angle for the second image.                                               |
| `speed?`                     | `number`                                                   | `1.6`                     | Speed of the animation (in seconds).                                                               |
| `easing?`                    | `Easing`                                                   | `easeOut`                 | `'easeIn' \| 'easeOut' \| 'easeInOut'`                                                             |
| `backgroundSize?`            | `BackgroundSize`                                           | `'cover'`                 | The background size of images, `'contain' \| 'cover' \| 'repeat' \| 'stretch'`                     |
| `displacmentBackgroundSize?` | `BackgroundSize`                                           | `'cover'`                 | The background size of displacement image, `'contain' \| 'cover' \| 'repeat' \| 'stretch'`         |
| `initialIndex?`              | `number`                                                   | `0`                       | The index of first image to display                                                                |
| `resizeDebounce?`            | `number`                                                   | `250`                     | Resize debounce in ms                                                                              |
| `dpr?`                       | `number`                                                   | `window.devicePixelRatio` | The pixel ratio of the canvas                                                                      |

### Return value

| Name           | Type                                                       | Description                                                   |
| -------------- | ---------------------------------------------------------- | ------------------------------------------------------------- |
| `next`         | `() => void`                                               | Transition to next image.                                     |
| `prev`         | `() => void`                                               | Transition to previous image.                                 |
| `jump`         | `(index: number) => void`                                  | Transition to specific image.                                 |
| `currentIndex` | `number`                                                   | The current index of carousel.                                |
| `ref`          | `MutableRefObject<T extends HTMLElement = HTMLDivElement>` | The ref of the parent DOM element.                            |
| `loadedImages` | `Partial<Record<number, boolean>>`                         | An object that represents the loaded images by their indexes. |


CHANGELOG.md
# [1.1.0](https://github.com/shasaf1986/distortion-effect-carousel/compare/v1.0.3...v1.1.0) (2020-05-07)


### Features

* **release:** deploy to npm ([a148e21](https://github.com/shasaf1986/distortion-effect-carousel/commit/a148e21abc9dae50115c9a225aba7acbff674171))

## [1.0.3](https://github.com/shasaf1986/distortion-effect-carousel/compare/v1.0.2...v1.0.3) (2020-05-05)


### Bug Fixes

* **release:** fix release ci ([a005373](https://github.com/shasaf1986/distortion-effect-carousel/commit/a0053730d9963f518c692167c52616d57fb1a930))

## [1.0.2](https://github.com/shasaf1986/distortion-effect-carousel/compare/v1.0.1...v1.0.2) (2020-05-05)


### Bug Fixes

* **release:** fix release ([7dd67cb](https://github.com/shasaf1986/distortion-effect-carousel/commit/7dd67cb367186fc475eb9cfe37eeb34f00eebfb7))

## [1.0.1](https://github.com/shasaf1986/distortion-effect-carousel/compare/v1.0.0...v1.0.1) (2020-05-05)


### Bug Fixes

* **fix release:** replaced package name ([c1a25fe](https://github.com/shasaf1986/distortion-effect-carousel/commit/c1a25fe0c18247adc2b321d05b6cd5f758b6c8da))

# 1.0.0 (2020-05-05)


### Features

* **release:** initial release ([7583cfa](https://github.com/shasaf1986/distortion-effect-carousel/commit/7583cfa6e2343031805b5c8b352b047aee5ff056))


src/distortionEffectCarousel.ts
import { TweenMax, Expo } from 'gsap';
import debounce from 'lodash/debounce';
import find from 'lodash/find';
import {
  ShaderMaterial,
  PlaneBufferGeometry,
  Mesh,
  Scene,
  OrthographicCamera,
  WebGLRenderer,
  Vector4,
  ImageBitmapLoader,
  Cache,
} from 'three';
import ResizeObserver from 'resize-observer-polyfill';
import { fragment, vertex } from './shaders';
import { TextureWrapper } from './textureWrapper';
import { cover, contain } from 'intrinsic-scale';
import './createImageBitmapPolyfill';
import { ImageWrapper } from './imageWrapper';

Cache.enabled = true;

export type BackgroundSize = 'cover' | 'contain' | 'repeat' | 'stretch';

export type Easing = keyof typeof Expo;

export type onImageLoaded = (index: number) => void;

export interface DistortionEffectCarouselPluginOptions {
  intensity?: number;
  commonAngle?: number;
  angle1?: number;
  angle2?: number;
  speed?: number;
  easing?: Easing;
  parent: HTMLElement;
  initialIndex?: number;
  backgroundSize?: BackgroundSize;
  displacmentBackgroundSize?: BackgroundSize;
  images: string[];
  displacmentImage: string;
  resizeDebounce?: number;
  onImageLoaded?: onImageLoaded;
  dpr?: number;
}

export class DistortionEffectCarouselPlugin {
  private readonly backgroundSize: BackgroundSize;

  private readonly displacmentBackgroundSize: BackgroundSize;

  private readonly images: ImageWrapper[];

  private displacmentImage: ImageWrapper;

  private readonly loader: ImageBitmapLoader;

  private readonly speed: number;

  private readonly easing: gsap.EaseFunction;

  private readonly parent: HTMLElement;

  private readonly scene: Scene;

  private readonly camera: OrthographicCamera;

  private readonly renderer: WebGLRenderer;

  private readonly mat: ShaderMaterial;

  private readonly geometry: PlaneBufferGeometry;

  private readonly texture1: TextureWrapper;

  private readonly texture2: TextureWrapper;

  private readonly disp: TextureWrapper;

  private readonly textures: TextureWrapper[];

  private readonly resizeObserver: ResizeObserver;

  private readonly angle1: number;

  private readonly angle2: number;

  private readonly dpr: number;

  private readonly onImageLoaded?: onImageLoaded;

  private isDisposed: boolean = false;

  private showFirstImage: boolean = true;

  private currentIndex: number = 0;

  private tween?: gsap.core.Tween;

  private width: number;

  private height: number;

  constructor({
    intensity = 1,
    commonAngle = Math.PI / 4,
    angle1,
    angle2,
    speed = 1.6,
    easing = 'easeOut',
    parent,
    initialIndex = 0,
    images,
    displacmentImage,
    backgroundSize = 'cover',
    displacmentBackgroundSize = 'cover',
    resizeDebounce = 250,
    onImageLoaded,
    dpr = (typeof window !== 'undefined' && window.devicePixelRatio) || 1,
  }: DistortionEffectCarouselPluginOptions) {
    this.dpr = dpr;
    this.onImageLoaded = onImageLoaded;
    this.backgroundSize = backgroundSize;
    this.displacmentBackgroundSize = displacmentBackgroundSize;
    this.loader = new ImageBitmapLoader();
    this.displacmentImage = new ImageWrapper({
      isDisplacement: true,
    });
    this.images = images.map(
      (_, index) => new ImageWrapper({ isDisplacement: false, index })
    );
    this.loader.crossOrigin = '';
    this.angle1 = typeof angle1 === 'number' ? angle1 : commonAngle;
    this.angle2 = typeof angle2 === 'number' ? angle2 : -commonAngle * 3;
    this.speed = speed;
    this.easing = Expo[easing];
    this.parent = parent;
    this.width = parent.offsetWidth;
    this.height = parent.offsetHeight;
    this.currentIndex = initialIndex;

    if (!this.checkIfIndexValid(this.currentIndex)) {
      console.warn('initial index out of boundaries');
      this.currentIndex = 0;
    }

    this.texture1 = new TextureWrapper({
      isDisplacement: false,
      index: this.currentIndex,
    });
    this.texture2 = new TextureWrapper({
      isDisplacement: false,
    });
    this.disp = new TextureWrapper({
      isDisplacement: true,
    });
    this.textures = [this.texture1, this.texture2, this.disp];

    this.mat = new ShaderMaterial({
      uniforms: {
        intensity: {
          type: 'f',
          value: intensity,
        },
        dispFactor: {
          type: 'f',
          value: 0.0,
        },
        angle1: {
          type: 'f',
        },
        angle2: {
          type: 'f',
        },
        texture1: {
          type: 't',
          value: this.texture1.texture,
        },
        texture2: {
          type: 't',
          value: this.texture2.texture,
        },
        disp: {
          type: 't',
          value: this.disp.texture,
        },
        res: {
          type: 'vec4',
          value: this.createVector(),
        },
      },
      vertexShader: vertex,
      fragmentShader: fragment,
      transparent: true,
      opacity: 1.0,
    });

    this.geometry = new PlaneBufferGeometry(this.width, this.height, 1);

    const mesh = new Mesh(this.geometry, this.mat);

    this.scene = new Scene();
    this.scene.add(mesh);

    this.camera = new OrthographicCamera(
      this.width / -2,
      this.width / 2,
      this.height / 2,
      this.height / -2,
      1,
      1000
    );

    this.camera.position.z = 1;

    this.renderer = new WebGLRenderer({
      alpha: true,
    });
    this.renderer.setPixelRatio(this.dpr);
    this.renderer.setClearColor(0xffffff, 0.0);
    this.renderer.setSize(this.width, this.height);
    this.renderer.compile(this.scene, this.camera);

    this.parent.appendChild(this.renderer.domElement);
    this.onResize = debounce(this.onResize, resizeDebounce);
    this.resizeObserver = new ResizeObserver(this.onResize);
    this.resizeObserver.observe(this.parent);
    this.loadImage(displacmentImage, this.displacmentImage);
    images.forEach((src, index) => {
      this.loadImage(src, this.images[index]);
    });
  }

  private loadImage(src: string, imageWrapper: ImageWrapper) {
    const { isDisplacement, index } = imageWrapper;
    this.loader.load(
      src,
      (image) => {
        if (this.isDisposed) {
          return;
        }
        imageWrapper.image = image;
        this.drawImage(imageWrapper);
        const textureWrapper = find(this.textures, { isDisplacement, index });
        if (textureWrapper) {
          this.attachImage(textureWrapper);
          if (!isDisplacement && index === this.currentIndex) {
            this.render();
          }
        }
        if (!isDisplacement && this.onImageLoaded) {
          this.onImageLoaded(index);
        }
      },
      () => undefined,
      () => {
        console.warn('faild to load image');
        if (!isDisplacement && this.onImageLoaded) {
          this.onImageLoaded(index);
        }
      }
    );
  }

  private render = () => {
    this.renderer.render(this.scene, this.camera);
  };

  private createVector() {
    return new Vector4(this.width, this.height, 1, 1);
  }

  private drawImage({ canvas, image, isDisplacement }: ImageWrapper) {
    if (!image) {
      return;
    }

    canvas.width = this.width * this.dpr;
    canvas.height = this.height * this.dpr;
    const context = canvas.getContext('2d');
    if (!context) {
      console.warn('canvas context is not defined');
      return;
    }
    context.clearRect(0, 0, canvas.width, canvas.height);
    const backgroundSize = isDisplacement
      ? this.displacmentBackgroundSize
      : this.backgroundSize;
    switch (backgroundSize) {
      case 'repeat': {
        const pattern = context.createPattern(image, 'repeat');
        context.fillStyle = pattern!;
        context.fillRect(0, 0, canvas.width, canvas.height);
        break;
      }
      case 'stretch': {
        context.drawImage(image, 0, 0, canvas.width, canvas.height);
        break;
      }
      default: {
        const scaleFunction = backgroundSize === 'cover' ? cover : contain;
        const { height, width, x, y } = scaleFunction(
          this.width,
          this.height,
          image.width,
          image.height
        );
        context.drawImage(
          image,
          x * this.dpr,
          y * this.dpr,
          width * this.dpr,
          height * this.dpr
        );
      }
    }
  }

  private attachImage(textureWrapper: TextureWrapper) {
    const { texture, index, isDisplacement } = textureWrapper;
    const { canvas } = isDisplacement
      ? this.displacmentImage
      : this.images[index];
    texture.image = canvas;
    texture.needsUpdate = true;
  }

  private checkIfIndexValid(index: number) {
    return index >= 0 && index <= this.images.length - 1;
  }

  private calculatePrevIndex(index: number) {
    let newIndex = index - 1;
    if (!this.checkIfIndexValid(newIndex)) {
      newIndex = this.images.length - 1;
    }

    return newIndex;
  }

  private calculateNextIndex(index: number) {
    let newIndex = index + 1;
    if (!this.checkIfIndexValid(newIndex)) {
      newIndex = 0;
    }

    return newIndex;
  }

  private onResize = () => {
    if (this.isDisposed) {
      return;
    }
    const newWidth = this.parent.offsetWidth;
    const newHeight = this.parent.offsetHeight;
    if (this.width === newWidth && this.height === newHeight) {
      return;
    }

    this.width = newWidth;
    this.height = newHeight;
    const { uniforms } = this.mat;
    this.drawImage(this.displacmentImage);
    this.images.forEach((image) => {
      this.drawImage(image);
    });
    this.textures.forEach(({ texture }) => {
      if (texture.image) {
        texture.needsUpdate = true;
      }
    });
    uniforms.res.value = this.createVector();

    this.renderer.setSize(this.width, this.height);
    this.render();
  };

  private slide(index: number, isNext: boolean) {
    this.currentIndex = index;
    const { uniforms } = this.mat;
    this.showFirstImage = !this.showFirstImage;
    const newTextureWrapper = this.showFirstImage
      ? this.texture1
      : this.texture2;
    newTextureWrapper.index = this.currentIndex;
    if ((this.showFirstImage && isNext) || (!this.showFirstImage && !isNext)) {
      uniforms.angle1.value = this.angle1;
      uniforms.angle2.value = this.angle2;
    } else {
      uniforms.angle1.value = this.angle2;
      uniforms.angle2.value = this.angle1;
    }

    this.attachImage(newTextureWrapper);
    this.tween?.kill();
    this.tween = TweenMax.to(uniforms.dispFactor, this.speed, {
      value: this.showFirstImage ? 0 : 1,
      ease: this.easing,
      onUpdate: this.render,
      onComplete: this.render,
    });
  }

  next() {
    const newIndex = this.calculateNextIndex(this.currentIndex);
    this.slide(newIndex, true);
  }

  prev() {
    const newIndex = this.calculatePrevIndex(this.currentIndex);
    this.slide(newIndex, false);
  }

  jump(index: number) {
    if (!this.checkIfIndexValid(index)) {
      console.warn('index out of boundaries');
      return;
    }
    if (index === this.currentIndex) {
      return;
    }

    const isNext = index > this.currentIndex;
    this.slide(index, isNext);
  }

  getCurrentIndex() {
    return this.currentIndex;
  }

  dispose() {
    this.tween?.kill();
    this.textures.forEach(({ texture }) => {
      texture.dispose();
    });
    this.parent.removeChild(this.renderer.domElement);
    this.renderer.dispose();
    this.scene.dispose();
    this.geometry.dispose();
    this.resizeObserver.disconnect();
    this.isDisposed = true;
  }
}
